src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -183,6 +121,136 @@ public GeneratorDriver AddAdditionalTexts(ImmutableArray<AdditionalText> additio
+                    try
+                    {
+                        generator.Initialize(context);
+                    }
+                    catch (Exception e)
+                    {
+                        ex = e;
+                    }
+                    generatorState = ex is null
+                                     ? new GeneratorState(context.InfoBuilder.ToImmutable())
+                                     : SetGeneratorException(MessageProvider, GeneratorState.Uninitialized, generator, ex, diagnosticsBag, isInit: true);
+                }
+
+                // create the syntax receiver if requested
+                if (generatorState.Info.SyntaxReceiverCreator is object)
+                {
+                    try
+                    {
+                        var rx = generatorState.Info.SyntaxReceiverCreator();
+                        walkerBuilder.SetItem(i, new GeneratorSyntaxWalker(rx));
+                        generatorState = generatorState.WithReceiver(rx);
+                        receiverCount++;
+                    }
+                    catch (Exception e)
+                    {
+                        generatorState = SetGeneratorException(MessageProvider, generatorState, generator, e, diagnosticsBag);
+                    }
+                }
+
+                stateBuilder.Add(generatorState);
+            }
+
+
+            // Run a syntax walk if any of the generators requested it
+            if (receiverCount > 0)
+            {
+                foreach (var tree in compilation.SyntaxTrees)
+                {
+                    var root = tree.GetRoot(cancellationToken);
+
+                    // https://github.com/dotnet/roslyn/issues/42629: should be possible to parallelize this
+                    for (int i = 0; i < walkerBuilder.Count; i++)
+                    {
+                        var walker = walkerBuilder[i];
+                        if (walker is object)
+                        {
+                            try
+                            {
+                                walker.Visit(root);
+                            }
+                            catch (Exception e)
+                            {
+                                stateBuilder[i] = SetGeneratorException(MessageProvider, stateBuilder[i], state.Generators[i], e, diagnosticsBag);
+                            }
+                        }
+                    }
+                }
+            }
+            walkerBuilder.Free();
+
+            // https://github.com/dotnet/roslyn/issues/42629: should be possible to parallelize this
+            for (int i = 0; i < state.Generators.Length; i++)
+            {
+                var generator = state.Generators[i];
+                var generatorState = stateBuilder[i];
+
+                // don't try and generate if initialization or syntax walk failed
+                if (!generatorState.Info.Initialized || generatorState.Exception is object)
```

Comment:
>!generatorState.Info.Initialized [](start = 20, length = 32)

Is this check necessary? Could we just`Debug.Assert(generatorState.Info.Initialized);` after checking `generatorState.Exception`?